test/framework/src/main/java/org/elasticsearch/common/util/concurrent/DeterministicTaskQueue.java

```diff
@@ -310,7 +315,11 @@ public String toString() {
 
             @Override
             public long relativeTimeInNanos() {
-                throw new AssertionError("DeterministicTaskQueue does not support nanosecond-precision timestamps");
+                if (nanoTimeSupplier != null) {
+                    return nanoTimeSupplier.getAsLong();
+                } else {
+                    return TimeValue.timeValueMillis(currentTimeMillis).nanos();
+                }
```

Comment:
It looks suspicious that deterministic task queue might rely on a time supplied from outside (what if it gets out of sync with relativeTimeInMillis?). Is it possible to always rely on `TimeValue.timeValueMillis(currentTimeMillis).nanos()`?

Reply:
Unfortunately it is not possible until #95800 is resolved.

If we were rely only on `TimeValue.timeValueMillis(currentTimeMillis).nanos()` there is a test SnapshotResiliencyTests#testSnapshotPrimaryRelocations which would become flaky.  In short, the test creates cluster with random number of master nodes and it works as expected iff there is only one master node is present. In case of multiple master nodes, the test simulates failure of master node to trigger new node election https://github.com/elastic/elasticsearch/blob/be78530dd28515634d57640d652a0a988e55cd17/server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java#L1011-L1013
This is done to test snapshot creating during master node failure.

When cluster state changes and unassigned info flows via the wire. Upon receiving UnassignedInfo object gets created via its constructor with *unassignedTimeNanos* field set to current nano time
https://github.com/elastic/elasticsearch/blob/51e1188304fec338d4472b4e7a57c57356dbbdbf/server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java#L298

Ultimately the test fails during *reroute* phase with assertion `assert nanoTimeNow >= unassignedTimeNanos` where *nanoTimeNow* comes from Deterministic task queue (and is artificial number which is not related to actual time flow) and *unassignedTimeNanos* is an actual nano time which is taken UnassignedInfo instance creation.

Bottom line 1
By providing *nanoTimeSupplier* in Deterministic task queue we are making *nanoTimeNow* and *unassignedTimeNanos* (from code above) mutually comparable since both of them are related to natural time flow for particular scenario.

Bottom line 2
Crux of #95800 is to stop relying on nano time for *RemainingDelay* calculation - the only one place where nano time is used in UnassignedInfo. When #95800 is fixed, we could remove *nanoTimeSupplier* from DeterministicTaskQueue and make code more cleaner.

Tip of stacktrace
```
java.lang.AssertionError
	at org.elasticsearch.cluster.routing.UnassignedInfo.getRemainingDelay(UnassignedInfo.java:448)
	at org.elasticsearch.cluster.routing.allocation.AllocationService.removeDelayMarkers(AllocationService.java:352)
	at org.elasticsearch.cluster.routing.allocation.AllocationService.reroute(AllocationService.java:529)
	at org.elasticsearch.cluster.routing.allocation.AllocationService.executeWithRoutingAllocation(AllocationService.java:458)
	at org.elasticsearch.cluster.routing.allocation.AllocationService.reroute(AllocationService.java:434)
	at org.elasticsearch.cluster.routing.allocation.AllocationService.disassociateDeadNodes(AllocationService.java:289)
	at org.elasticsearch.cluster.coordination.NodeLeftExecutor.execute(NodeLeftExecutor.java:88)
	at org.elasticsearch.cluster.service.MasterService.innerExecuteTasks(MasterService.java:1039)
	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:1004)
	at org.elasticsearch.cluster.service.MasterService.executeAndPublishBatch(MasterService.java:232)
	at org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.lambda$run$2(MasterService.java:1622)
```